Lib/idlelib/pyparse.py

```diff
@@ -101,46 +101,19 @@
 """, re.VERBOSE).match
 
 
-class StringTranslatePseudoMapping(Mapping):
-    r"""Utility class to be used with str.translate()
+class ParseMap(dict):
+    r"""Dict subclass that maps anything not in dict to 'x'.
 
-    This Mapping class wraps a given dict. When a value for a key is
-    requested via __getitem__() or get(), the key is looked up in the
-    given dict. If found there, the value from the dict is returned.
-    Otherwise, the default value given upon initialization is returned.
+    This is designed to be used with str.translate in study1.
+    Anything not specifically mapped otherwise becomes 'x'.
+    Example: replace everything except whitespace with 'x'.
 
-    This allows using str.translate() to make some replacements, and to
-    replace all characters for which no replacement was specified with
-    a given character instead of leaving them as-is.
-
-    For example, to replace everything except whitespace with 'x':
-
-    >>> whitespace_chars = ' \t\n\r'
-    >>> preserve_dict = {ord(c): ord(c) for c in whitespace_chars}
-    >>> mapping = StringTranslatePseudoMapping(preserve_dict, ord('x'))
-    >>> text = "a + b\tc\nd"
-    >>> text.translate(mapping)
+    >>> keepwhite = ParseMap((ord(c),ord(c)) for c in ' \t\n\r')
+    >>> "a + b\tc\nd".translate(keepwhite)
     'x x x\tx\nx'
     """
-    def __init__(self, non_defaults, default_value):
-        self._non_defaults = non_defaults
-        self._default_value = default_value
-
-        def _get(key, _get=non_defaults.get, _default=default_value):
-            return _get(key, _default)
-        self._get = _get
-
-    def __getitem__(self, item):
-        return self._get(item)
-
-    def __len__(self):
-        return len(self._non_defaults)
-
-    def __iter__(self):
-        return iter(self._non_defaults)
-
-    def get(self, key, default=None):
-        return self._get(key)
+    def __getitem__(self, key):
+        return self.get(key, 'x')
```

Comment:
I had done a version yesterday that inherited from `dict` and only implemented:
```
class Parse(dict):
    def __missing__(self, key):
        return 'x'
```
It didn't have `get` or `__getitem__` functionality on missing values, but it worked for the translate.  But then I read on SO that's it's generally not good to inherit directly from `dict` and to use `Mapping` or `MutableMapping` instead.  I guess this doesn't matter here, but I did notice that inheriting from `dict` allows some attributes to change from read-only to not, such as `_tran.update = {1: 2}`.  So when is it OK to inherit directly from dict?

Reply:
When it works.  *We* are not going to do anything so nonsensical.

Reply:
:-)

Reply:
Defining `__missing__` looks the most natural and efficient way. This is the case for which `__missing__` was introduced.